File: Leetcode/2130. Maximum Twin Sum of a Linked List.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    int pairSum(ListNode* head) {
        int maxSum = 0;
        ListNode* slowPtr = head, *fastPtr = head -> next;
        stack <int> Stack;
        Stack.push(slowPtr -> val);
        while (fastPtr -> next) {
            slowPtr = slowPtr -> next;
            Stack.push(slowPtr -> val);
            fastPtr = fastPtr -> next -> next;
        }
        slowPtr = slowPtr -> next;
        while (!Stack.empty()) {
            maxSum = max(maxSum, Stack.top() + slowPtr -> val);
            slowPtr = slowPtr -> next;
            Stack.pop();
        }
        return maxSum;
    }
};

File: Leetcode/2130. Maximum Twin Sum of a Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def pairSum(self, head: Optional[ListNode]) -> int:
        slowPtr, fastPtr, Stack, maxSum = head, head.next, [head.val], 0
        while fastPtr.next:
            slowPtr = slowPtr.next
            Stack.append(slowPtr.val)
            fastPtr = fastPtr.next.next
        slowPtr = slowPtr.next
        while len(Stack):
            maxSum = max(maxSum, slowPtr.val + Stack[-1])
            Stack.pop()
            slowPtr = slowPtr.next
        return maxSum